### PR TITLE
fix: supply argv to config with functions

### PR DESCRIPTION
**Pull request by anshumanv**
**What kind of change does this PR introduce?**

fix

**Did you add tests for your changes?**
Yes

**If relevant, did you update the documentation?**
NA

**Summary**

- Pass argv to config function

**Does this PR introduce a breaking change?**
No

**Other information**
Fixes https://github.com/webpack/webpack-cli/issues/1704
Fixes https://github.com/webpack/webpack-cli/issues/1126

### CODE REVIEW

File: test/config/type/function-with-argv/function-with-argv.test.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+const { existsSync } = require('fs');
+const { resolve } = require('path');
+const { run } = require('../../../utils/test-utils');
+
+describe('function configuration', () => {
+    it('is able to understand a configuration file as a function', () => {
+        const { stderr, stdout } = run(__dirname, ['--mode', 'development']);
+        expect(stderr).toBeFalsy();
+        expect(stdout).toBeTruthy();
+        // Should generate the appropriate files
```
Comment (anshumanv) [inline test/config/type/function-with-argv/function-with-argv.test.js:11]: @evilebottnawi any way to make these tests better?

Comment (snitin315) [reply test/config/type/function-with-argv/function-with-argv.test.js:11]: is it possible to log `argv` and check its value? It should not be `undefined`

Comment (evilebottnawi) [reply test/config/type/function-with-argv/function-with-argv.test.js:11]: I think testing is enough in this case

Comment (anshumanv) [reply test/config/type/function-with-argv/function-with-argv.test.js:12]: oops I added a commit for checking argv too, I guess it's fine

**Comment by milizhang** [comment]
I think this does not actually work - while it seems like the arguments now can be passed to config function, when you actually tries to pass custom arguments, user will stuck at "Which flags do you want to use", likely as the custom argument is unknown to the cli.

**Comment by anshumanv** [comment]
@milizhang can you give example?

**Comment by milizhang** [comment]
@anshumanv Please take a look at this: https://github.com/milizhang/webpack-cli-4-argv-repro

If you use webpack-cli 3, `console.log(argv.testArg);` will give you "Test" if you run `webpack --testArg=Test`, while webpack-cli 4 will stuck asking you to clarify which argument you want to place. (Unless it is an argument that already exists in core webpack).

**Comment by anshumanv** [comment]
Yep unknown arguments are not registered, you should use `--env` for it

**Comment by milizhang** [comment]
Is this something that webpack-cli explicitly decided to deprecate in V4? Are there anyway to register custom arguments in configurations?

**Comment by anshumanv** [comment]
> Is this something that webpack-cli explicitly decided to deprecate in V4? Are there anyway to register custom arguments in configurations?

Yes, passing arbitrary values was deprecated. You should be able to achieve it using `--env` or does it not fit your use case?

You can just do - 

`--env testArg=test`

```js
env = { testArg: 'test' } // in config 
```

**Comment by milizhang** [comment]
Got it. Picking custom argv vs. env was more of a preference issue. Since it is deprecated I will move toward env.

Thanks for the info.

**Pull request merged**